### ISSUE TITLE
Top 10 Medium Term Project Priorities

### ISSUE THREAD

**Issue by marcushyett-ph** [PostHog/posthog#9514]
This is a living list of project level priorities across product teams, to ensure have the context we need to solve these problems and are always prioritizing the most impactful work during our sprints. It does not include items we expect to finish in the current sprint.

"Nail Enterprise" by (company goals):
* **Quality:** Our core features (insights, recordings, feature-flags / experimentation) work like a Swiss watch
* **Extensibility:** You can instantly connect PostHog to any other system
* **Deployments:** PostHog is the easiest self-hosted product to deploy and scale in the world

They are in priority order, please push back if you disagree with the priorities. Where you see context TBD - I'm still gathering details for this - if you already have context feel free to add

# Top 10

1. **Persons on events** - enabling significantly higher scale and more accurate querying by migrating persons to events table (Context: TBD)
    * **Why urgent**: This is an immediate blocker to onboarding very large clients, and a dependency for large query performance improvements using caching
    * **Why important**: Scale is a significant differentiator for us and enables us to open up a new market for large self-hosted enterprise customers
    * **Status:** In Progress
1. **Standardize infrastructure** - our cloud infrastructure today is a snowflake, we do not use the same self-hosted tooling we offer others (Context: TBD)
    * **Why urgent**: We've experienced a number of fires recently which better standardization of infrastructure could have prevented
    * **Why important**: Increasing reliability and simplicity of our cloud infrastructure will free up resources to put more energy into proactively improving the self-hosting experience and scalability
    * **Status:** In Progress
1. **Onboarding (Iterative experiments)**: The onboarding process to PostHog is difficult, leading us to loose potentially valuable customers each day, including enterprise customers (Context: TBD)
    * **Why urgent** We're losing out on potentially 100s of potentially valuable customers who could be successful with PostHog each week, because of the difficulty of getting data into PostHog
    * **Why important?** Any improvement to onboarding now will have a long term affect on the success as we will have more happy customers which will drive more word of mouth
    * **Status:** In Progress
1. **Subscription conversion** On cloud and self-hosting we are not effective in enabling people to find, understand the value of and convert to subscribed features (Context: TBD)
    * **Why urgent:** The sooner we do this we have a higher potential to capture more revenue and retain more users by enabling them to access valuable extra features
    * **Why important:** Improving the subscription conversion rate will directly improve revenue in the long term
    * **Status:** Design & Product Context Gathering
1. **Multivariate Feature Flag support for top libraries** - Today we only support Multivariate feature flags on limited libraries, this means many of our biggest customers have to use workarounds for experimentation and feature flags (Context:[ Popularity of libraries by team](https://metabase.posthog.net/question/262-client-library-usage))
    * **Why urgent:** A number of large clients are currently working around this limitation using custom implementation for experiments, this will be hard to maintain
    * **Why important:** We're limiting adoption of experimentation and mutivariate feature flags, as well as overall adoption of PostHog by offering limited library support
    * **Status:** Ready to implement (no design or additional product context required)
1. **Session analysis**: PostHog lacks basic session analysis functionality, making it difficult to measure user engagement with a product (Context: [Product Context](https://docs.google.com/document/d/1rj0yMbxwR_BYCTJNct4x-iCn6yv2mJfNy28OV6bxkAA/edit?usp=sharing))
    * **Why urgent:** Customers have been frequently asking to bring back reliable session measurement functionality, it blocks them from moving away from Google Analytics
    * **Why important:** Measuring engagement is key for our customers to understand user behavior and improve the product for the most and least engaged users, many customers cannot use session recording and have no alternative method in PostHog
    * **Status:** Product Context Gathering
1. **App Platform (Iteration 1)**: Our plugins ecosystem has driven some initial traction however, there are a number of frequently requested features which cannot be built on top of this ecosystem (Context: [product context doc] (https://docs.google.com/document/d/1IbxrazOKYidQQjqxJ8phJ-XMi1IeRRikbT3g0dwgkOc/edit#heading=h.fi42es92skqe))
    * **Why urgent:** Following the hackathon, there is a lot of momentum around launching "apps" including improvements to existing plugin documentation (with significant work done already from the marketing side)
    * **Why important:**: Extensibility is a key priority for the company to nail enterprise, supporting more sophisticated apps will enable our customers to extend PostHog for their specific needs with less overhead
    * **Status:** Design and eng PoC in progress
1. **Simplifying events and actions**: Events and actions are very separate concepts which cause confusion and make it difficult to  create good data management (Context: [Product context](https://github.com/PostHog/product-internal/pull/300))
    * **Why urgent:** Most large customers signing up are confused between events and action making it harder to get started and maintain consistency across their organization
    * **Why important:** Data management is critical for enterprise customers to be successful, the separate concepts of actions and events causes confusion and inefficiency in getting insights from PostHog
    * **Status:** Technical and Experience Design
1. **UI Component reusability** - Make a subset of UI components easy to reuse across the product (Context: TBD)
    * **Why urgent**: Increasing quality is critical for nailing enterprise, anything we can do to reduce the effort required to increase and maintain quality of the experience will have a exponential impact over time
    * **Why important**: The quality of our product is key to gain and retain enterprise clients, inconsistencies and reliability issues will cause our customers to lose trust
    * **Status:** Initial Design Explorations
1. **Export visualizations** People share their findings from PostHog regularly via links or screenshots, the results are often inconsistent and go out-of-date quickly (Context: TBD)
    * **Why urgent:** [There's a steady increase in organizations with multiple users of the past months](https://app.posthog.com/insights/qyMSn6Nf), sharing outside of is key to growing overall engagement particularly with larger teams
    * **Why important:** Results of analytics should inform every area of a company, from leadership to marketing, product and engineering, making it easy to share specific findings in a format people are familiar with enables people to get more value out of PostHog
    * **Status:** Product Context Gathering

------ 
Next Up after Top 10
------
 1. **Insight & Dashboard subscriptions** - People don't have time  to visit PostHog and view dashboards every day, we can make this easier to consume this important data by sending them regular updates, to where they are (Context: TBD)
    * **Why urgent:** Blocked by Export visualizations, this will help teams (including our own keep on top of our key metrics)
    * **Why important:** This will reduce the barriers to getting quick regular value from PostHog for people who are interested in the same set of things, helping us grow engagement particularly in larger organizations
    * **Status:**  Product Context Gathering
1. **Universal querying** - enabling advanced querying across all areas of the platform, removing barriers between features and insights (Context: TBD)
    * **Why urgent:** We have put significant work into improving cohort querying, yet this same level of querying will not be available within other areas of PostHog
    * **Why important:** Two key aspects of quality are completeness and consistency, enabling a complete querying experience (such as in new cohorts) across every insight and entity in PostHog will improve quality and make discovering new insights easier and faster
    * Status: Product Context Gathering (initial PoC developed in hackathon)

**Comment by tiina303** [comment]
What about: 
Making self-hosted deployments easier and more reliable (since we already mention Standardize infrastructure, which I'm not sure is a priority across product teams), e.g.
     - monitoring, alerting etc comes out of the box for enterprise customers
     - plugin-server doesn't just randomly stall and stop processing events and no-one knows about it nor why
     - Also our monitoring/alerting 

Question about Onboarding - does that cover the idea of making PostHog as easy to learn as a video game? Specifically how far is the onboarding project going? does it cover directing users to check out session recordings and later paths?

**Comment by marcushyett-ph** [comment]
> Making self-hosted deployments easier and more reliable (since we already mention Standardize infrastructure, which I'm not sure is a priority across product teams), e.g.
     - monitoring, alerting etc comes out of the box for enterprise customers
     - plugin-server doesn't just randomly stall and stop processing events and no-one knows about it nor why
     - Also our monitoring/alerting 

I think this should be covered by standardize infrastructure, but I agree its not clear as it is, we should either break this down into smaller discrete projects and prioritize them separately or document clearly what's in the scope. My intuition is to break it down further.

> Question about Onboarding - does that cover the idea of making PostHog as easy to learn as a video game? Specifically how far is the onboarding project going? does it cover directing users to check out session recordings and later paths?

We're planning to iteratively improve onboarding starting with the very basics like getting ingestion setup, we'll keep iterating and learning from experiments. As easy as a video game is the 3 year goal - so it'll be a while before we get there. @liyiy should be able to provide further context on where we're currently focussed.

**Comment by neilkakkar** [comment]
I'm a bit bummed to not see extending experimentation and multivariate support to all client libraries here. People like using experimentation / using multivariates FF but are constrained by our library support, since not everyone uses posthog-js, including some huuuge customers.

I would put this above number 5, sessions, given that polishing existing products > new features

**Comment by marcushyett-ph** [comment]
@neilkakkar great feedback, thank you!

I originally had this in the list but I removed it based on a slightly unrelated chat with @timgl around whether getting our libraries up to spec for (feature flags, experimentation, sessionIDs and other important inconsistencies) should be prioritized as kind-of tech debt priority (in parallel to this).

Happy to add it back in as a priority in this though - to make sure we have a single place for priorities?

**Comment by neilkakkar** [comment]
Yep, I'd like us to have only one priority list.

Also, I wouldn't call it tech-debt. It's more like this new feature has only been partially rolled out, which is harming things as people get excited about experiments -> try creating one -> realise its only posthog-js -> go away 😢. (and also additional burden on us to maintain docs about where it works and where it doesn't, and what to do about it)

Calling it tech debt encourages pushing it down the line, separates from feature-priority-list, which is possibly how we got into this situation that multivariates haven't been rolled out to libraries other than posthog-js for the past 10 months :/

**Comment by marcushyett-ph** [comment]
Cool - added it in. in my mind I was grouping together "Get top libraries up-to-date for FF, Sessions, etc etc", as one big tech  debt project. But It's easier to make progress if we scope it down to something specific.

> people get excited about experiments -> try creating one -> realise its only posthog-js -> go away 

I agree with this statement, but its definitely worth adding the context of [how many people use each library](https://metabase.posthog.net/question/262-client-library-usage), since js (web) is a lot more popular.

Also let's scope out (separate from this) the libraries we want to prioritize (if not all).

**Comment by liyiy** [comment]
@tiina303 We're currently focused on Onboarding 1.0 and 1.5 right after that. Videogamifying things should come in Onboarding 2.0 :D 

You can check out the onboarding doc here https://docs.google.com/document/d/1m1vmXlyM3rbz9GziJUV40SNJstpq6e15pKvUZfqBUB4/edit?usp=sharing for more details. 

Basically with 1.0 and 1.5 we are focusing right now on increasing conversion rates from signup -> first event ingested.

**Comment by clarkus** [comment]
> UI Component reusability - Make a subset of UI components easy to reuse across the product (Context: TBD)

This is in-progress, but it's been a lower priority. Most of the documentation so far focuses on design constants like color, type, spacing, etc. I'm starting to get into more form components now. 
https://www.figma.com/file/Y9G24U4r04nEjIDGIEGuKI/?node-id=2028%3A808

**Comment by mariusandra** [comment]
Great list! Really happy to see all these projects getting prioritised.

Here's something for the "Next Up after Top 10" list to get us excited about what's possible. Priorities to be determined.

**App platform (iteration 2)** (dependent on "App platform (iteration 1)")
- Users are able to build **any SaaS app** on top of PostHog. New slogan: *"the only app you self-host"*.
- PostHog provides the database for both persistent and event/timeseries-based data
- PostHog provides the build tooling, test suite, component library, and so on.
- Users are able to build all of this in the in-app code editor, which is basically four textareas: "config json", "frontend typescript source", "backend typescript source" and "client side typescript source".

**App platform (iteration 3)** (dependent on "App platform (iteration 2)")
- Users are able to build any plugin using an integrated vscode-like interface within posthog
- Git integration, autocomplete, etc.
- Users are able to test their backend code and run tests live in the interface.

**Bring your own ~Plugin~ App Server**
- Users are able to write their own apps on posthog cloud
- For this they either supply their own plugin/app server, which will handle ingestion and communicates with us through kafka (events in/out).
- On cloud, we can then spin up isolated app servers to separate paying users from free trial users and their hacky apps

**EU cloud** (dependent on "Standardize infrastructure")
- We're losing money here.

**Comment by rcmarron** [comment]
I love this list! Thanks for putting it together!

There's a set of projects related to session recordings that I think we should tackle after the session work listed above. I'm not sure how the priority should stack up, but would love to have a discussion around how the following should stack up:

1. **Session recordings stored outside Clickhouse**: We store recordings in ClickHouse today, but it’s not the ideal storage location for this data
    * **Why urgent**: Customer’s Clickhouse storage can fill up based on a front end change - breaking PostHog. This has happened to several customers already, and it’s a pretty terrible experience.
    * **Why important**: The reason above, and by completing this, we could expand recording retention significantly - which enables a set of features around sharing and interacting with recordings.
    * **Status**: Hackathon-ed
2. **Seamless session recording setup**: There are a set of session recording configurations that we have users hack around, but we should make them easy configs on the setting page. Specifically, limiting recordings to specific urls, only record when the user is “focused” on the screen, only record a percentage of users, and masking text input. Additionally, we should fix the confusion around recordings + permitted domains, and make it more obvious to users if they’re running two instances of posthog-js at once.
    * **Why urgent**: We get a good amount of recording support requests that are related to these setup issues. We give users work arounds, but recording setup should be dead simple.
    * **Why important**: We want more people being successful in setting up recordings.
    * **Status**: Product Context Gathering
3. **Charge for recordings**: Right now recordings are free, we should probably charge for them.
    * **Why urgent**: Users are using recordings for free right now. It’s going to be painful to switch this to a paid model. We should do it sooner rather than later.
    * **Why important:** Because we’re a business, and recordings cost us money, so we should charge for them
    * **Status**: Business side of things was done in prep for launch week (I think?)
4. **Recording playlist**: We should have a much better experience for users who want to view multiple recordings in sequence.
    * **Why urgent**: Most users watch recordings in batches, but it’s really painful to go back and forth between the list view and the recordings.
    * **Why important**: To get value from recordings, you need to watch a bunch of them. User’s are searching for “needles in a haystack”. Let’s make sure they’re finding the needles
    * **Status**: Product Context Gathering
5. **Recording debugging experience**: Users use recordings to debug customer issues. Viewing console logs and network requests can really help this experience.
    * **Why urgent**:  These are commonly requested features for our recordings product. Users are already trying to use the product in this way, but we don’t do a great job of supporting it
    * **Why important**: We want users to get as much value as possible from our recordings feature. Allowing for a debugging experience here feels like relatively low hanging fruit to really enhance the product experience.
    * **Status**: Console logs feature is in beta + needs to work - needs more product context gathering
